README.rst

```diff
@@ -0,0 +1,806 @@
+Context
+=======
+
+We build and release software by massively consuming and producing software
+packages such as NPMs, RPMs, Rubygems, etc.
+
+Each package manager, platform, type or ecosystem has its own conventions and
+protocols to identify, locate and provision software packages.
+
+
+Problem
+=======
+
+When tools, APIs and databases process or store multiple package types, it is
+difficult to reference the same software package across tools in a uniform way.
```

Comment:
Ok… so is the goal to provide a URI that:

1) That tells you which tool consumes this
AND
2) Provides enough information for that tool to consume it

?

Or is the goal only to provide a way to uniquely reference a specific package, not to be able to reconstruct how its installed?

Reply:
@iarna you wrote:

> Ok… so is the goal to provide a URI that:

>    That tells you which tool consumes this
>    AND
>    Provides enough information for that tool to consume it ?

Not exactly `which tools consumes it`: for instance a `maven` `purl` does not tell you to use Maven, or Gradle, SBT or Ivy,  an `npm` `purl` does not tell you you should use the `npm` tool over `yarn` or anything else. `pypi` does not tell you to use `pip` , `easy_install` or `buildout` .... what `npm` mor `pypi` mean here is the whole spet of packaging format, manifest, registry, protocols and APIs that a tool uses to deal with such package `type`.

And provide just enough to consume it: well that's a possibility though it may be just a side goal.

In #5 I toyed with the hypothetical idea of a "meta" package manager.... Though frankly this would be a just a franken manager IMHO.


> Or is the goal only to provide a way to uniquely reference a specific package, not to be able to reconstruct how its installed?

> Who would consume these? Humans? Software? Which software?

I think that the distinguished [racketist](https://racket-lang.org/) @jackfirth provides a better summary in #6 that I could have ever written. I intend to steal his words and add them to the spec:

> [...] here are the proposed use cases as I understand them:

> - Cross-system metadata indexing to search and monitor packages by metadata like available versions, dependencies, contributors, etc. across multiple package managers (libraries.io)

> - Vulnerability tracking to determine whether a package's set of possible transitive dependencies includes a known vulnerability and whether the version constraints of that dependency graph allow or prevent patching

> - Other kinds of package-content-agnostic analysis tools, especially tools that look at the dependency graphs of package ecosystems


So IMHO the primary beneficiaries are folks or DBs or APIs or tools that deal with several package formats and languages. Very selfishly,that might include me, @andrew @jayfk @jasinner @ashcrow @R2wenD2 @tgamblin @jpopelka and hopefully many more. This may be used in these context for UI and DB and APIs. 

It may also --if this gets enough traction-- be something that influences positively some standardization in the domain in the future as a side benefit: say tomorrow you decide to include an `npm` package size as a key identifying package attribute: may you will think about whether this makes sense purl-wise? and either reconsider this or contribute to `purl` to adopt such a fine new standard?